pyanaconda/payload/manager.py

```diff
@@ -0,0 +1,238 @@
+# Class for management payload threading.
+#
+# Copyright (C) 2019  Red Hat, Inc.
+#
+# This copyrighted material is made available to anyone wishing to use,
+# modify, copy, or redistribute it subject to the terms and conditions of
+# the GNU General Public License v.2, or (at your option) any later version.
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY expressed or implied, including the implied warranties of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General
+# Public License for more details.  You should have received a copy of the
+# GNU General Public License along with this program; if not, write to the
+# Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+# 02110-1301, USA.  Any Red Hat trademarks that are incorporated in the
+# source code or documentation are not subject to the GNU General Public
+# License and may only be used or replicated with the express permission of
+# Red Hat, Inc.
+#
+
+import threading
+
+from pyanaconda.core.constants import THREAD_STORAGE, THREAD_PAYLOAD, THREAD_PAYLOAD_RESTART, \
+    THREAD_WAIT_FOR_CONNECTING_NM
+from pyanaconda.core.i18n import _, N_
+from pyanaconda.threading import threadMgr, AnacondaThread
+from pyanaconda.payload import PackagePayload
+from pyanaconda.payload.errors import PayloadError
+
+from pyanaconda.anaconda_loggers import get_module_logger
+
+log = get_module_logger(__name__)
+
+
+__all__ = ["payloadMgr"]
+
+
+class PayloadManager(object):
+    """Framework for starting and watching the payload thread.
+
+    This class defines several states, and events can be triggered upon
+    reaching a state. Depending on whether a state has already been reached
+    when a listener is added, the event code may be run in either the
+    calling thread or the payload thread. The event code will block the
+    payload thread regardless, so try not to run anything that takes a long
+    time.
+
+    All states except STATE_ERROR are expected to happen linearly, and adding
+    a listener for a state that has already been reached or passed will
+    immediately trigger that listener. For example, if the payload thread is
+    currently in STATE_GROUP_MD, adding a listener for STATE_NETWORK will
+    immediately run the code being added for STATE_NETWORK.
+
+    The payload thread data should be accessed using the payloadMgr object,
+    and the running thread can be accessed using threadMgr with the
+    THREAD_PAYLOAD constant, if you need to wait for it or something. The
+    thread should be started using payloadMgr.restartThread.
+    """
+
+    STATE_START = 0
+    # Waiting on storage
+    STATE_STORAGE = 1
+    # Waiting on network
+    STATE_NETWORK = 2
+    # Verify repository availability
+    STATE_TEST_AVAILABILITY = 3
+    # Downloading package metadata
+    STATE_PACKAGE_MD = 4
+    # Downloading group metadata
+    STATE_GROUP_MD = 5
+    # All done
+    STATE_FINISHED = 6
```

Comment:
I would like to see these states removed or replaced with an enum. The numbers complicate debugging.

Reply:
I agree.

Reply:
Fixed.